File: src_python/rag/vector_store.py
"""
Vector Store Module
Manages storage and retrieval of document embeddings.
"""

import json
import uuid
from pathlib import Path
from typing import Dict, List, Any, Optional
from datetime import datetime

import numpy as np

from ..utils.logger import get_logger
from ..utils.config import RAGConfig


logger = get_logger("vector-store")


class VectorStore:
    """
    Store for document embeddings with search capabilities.
    """
    
    def __init__(self, config: RAGConfig):
        """
        Initialize vector store.
        
        Args:
            config: RAG configuration
        """
        self.config = config
        self.store_path = Path(config.vector_store_path) / "vectors"
        self.index_path = Path(config.vector_store_path) / "index.json"
        self.max_documents = config.max_documents
        self.index: Dict[str, Dict[str, Any]] = {}
    
    async def initialize(self) -> None:
        """
        Initialize vector store.
        """
        try:
            self.store_path.mkdir(parents=True, exist_ok=True)
            await self.load_index()
            logger.info("Vector store initialized")
        except Exception as error:
            logger.error("Error initializing vector store", {"error": str(error)})
            raise
    
    async def store(
        self,
        chunks: List[Dict[str, Any]]
    ) -> Dict[str, Any]:
        """
        Store document chunks with embeddings.
        
        Args:
            chunks: Chunks with embeddings to store
            
        Returns:
            Storage result with document ID
        """
        try:
            await self.initialize()
            
            document_id = f"doc-{datetime.now().timestamp()}-{uuid.uuid4().hex[:8]}"
            doc_path = self.store_path / f"{document_id}.json"
            
            document = {
                "id": document_id,
                "chunks": [
                    {
                        "id": chunk["id"],
                        "content": chunk["content"],
                        "embedding": chunk["embedding"],
                        "metadata": chunk.get("metadata", {})
                    }
                    for chunk in chunks
                ],
                "timestamp": datetime.now().isoformat(),
                "chunk_count": len(chunks)
            }
            
            # Save document to disk
            with open(doc_path, 'w', encoding='utf-8') as f:
                json.dump(document, f, ensure_ascii=False, indent=2)
            
            # Update index
            self.index[document_id] = {
                "id": document_id,
                "path": str(doc_path),
                "timestamp": document["timestamp"],
                "chunk_count": document["chunk_count"],
                "source": chunks[0].get("source", "") if chunks else ""
            }
            
            await self.save_index()
            
            logger.info("Document stored successfully", {
                "document_id": document_id,
                "chunk_count": chunks
            })
            
            return {"id": document_id, "chunk_count": len(chunks)}
            
        except Exception as error:
            logger.error("Error storing document", {"error": str(error)})
            raise
    
    async def search(
        self,
        query_embedding: List[float],
        top_k: int = 5,
        filters: Optional[Dict[str, Any]] = None
    ) -> List[Dict[str, Any]]:
        """
        Search for similar documents.
        
        Args:
            query_embedding: Query embedding vector
            top_k: Number of results to return
            filters: Optional filters for search
            
        Returns:
            List of similar chunks with similarity scores
        """
        try:
            await self.initialize()
            
            results = []
            document_ids = list(self.index.keys())
            
            for doc_id in document_ids:
                doc = await self.load_document(doc_id)
                if not doc:
                    continue
                
                for chunk in doc["chunks"]:
                    # Apply filters
                    if filters:
                        if "type" in filters and chunk["metadata"].get("type") != filters["type"]:
                            continue
                        if "source" in filters and chunk["metadata"].get("source") != filters["source"]:
                            continue
                    
                    # Calculate similarity
                    similarity = self._calculate_cosine_similarity(
                        query_embedding,
                        chunk["embedding"]
                    )
                    
                    results.append({
                        "document_id": doc_id,
                        "chunk_id": chunk["id"],
                        "content": chunk["content"],
                        "metadata": chunk["metadata"],
                        "similarity": similarity
                    })
            
            # Sort by similarity and return top K
            results.sort(key=lambda x: x["similarity"], reverse=True)
            return results[:top_k]
            
        except Exception as error:
            logger.error("Error searching vector store", {"error": str(error)})
            raise
    
    async def load_document(self, document_id: str) -> Optional[Dict[str, Any]]:
        """
        Load document from disk.
        
        Args:
            document_id: Document ID to load
            
        Returns:
            Document data or None if not found
        """
        try:
            doc_info = self.index.get(document_id)
            if not doc_info:
                logger.warning(f"Document not found in index: {document_id}")
                return None
            
            with open(doc_info["path"], 'r', encoding='utf-8') as f:
                return json.load(f)
        except Exception as error:
            logger.error("Error loading document", {
                "document_id": document_id,
                "error": str(error)
            })
            return None
    
    async def delete_document(self, document_id: str) -> bool:
        """
        Delete document from store.
        
        Args:
            document_id: Document ID to delete
            
        Returns:
            True if deleted successfully
        """
        try:
            doc_info = self.index.get(document_id)
            if not doc_info:
                logger.warning(f"Document not found: {document_id}")
                return False
            
            # Delete file
            Path(doc_info["path"]).unlink()
            
            # Remove from index
            del self.index[document_id]
            await self.save_index()
            
            logger.info(f"Document deleted: {document_id}")
            return True
            
        except Exception as error:
            logger.error("Error deleting document", {
                "document_id": document_id,
                "error": str(error)
            })
            raise
    
    async def save_index(self) -> None:
        """
        Save index to disk.
        """
        try:
            with open(self.index_path, 'w', encoding='utf-8') as f:
                json.dump(self.index, f, ensure_ascii=False, indent=2)
        except Exception as error:
            logger.error("Error saving index", {"error": str(error)})
            raise
    
    async def load_index(self) -> None:
        """
        Load index from disk.
        """
        try:
            if self.index_path.exists():
                with open(self.index_path, 'r', encoding='utf-8') as f:
                    self.index = json.load(f)
                logger.info(f"Loaded {len(self.index)} documents from index")
            else:
                # Index doesn't exist yet, create empty index
                self.index = {}
                await self.save_index()
        except Exception:
            self.index = {}
            await self.save_index()
    
    async def get_stats(self) -> Dict[str, Any]:
        """
        Get store statistics.
        
        Returns:
            Statistics dictionary
        """
        await self.initialize()
        
        document_ids = list(self.index.keys())
        total_chunks = sum(
            self.index[doc_id]["chunk_count"]
            for doc_id in document_ids
        )
        
        return {
            "total_documents": len(document_ids),
            "total_chunks": total_chunks,
            "max_documents": self.max_documents,
            "available_space": self.max_documents - len(document_ids)
        }
    
    async def clear(self) -> None:
        """
        Clear all documents from store.
        """
        try:
            document_ids = list(self.index.keys())
            
            for doc_id in document_ids:
                await self.delete_document(doc_id)
            
            logger.info("Vector store cleared")
        except Exception as error:
            logger.error("Error clearing vector store", {"error": str(error)})
            raise
    
    def _calculate_cosine_similarity(
        self,
        vec1: List[float],
        vec2: List[float]
    ) -> float:
        """
        Calculate cosine similarity between two vectors.
        
        Args:
            vec1: First vector
            vec2: Second vector
            
        Returns:
            Similarity score
        """
        if len(vec1) != len(vec2):
            raise ValueError("Vector dimensions must match")
        
        dot_product = sum(a * b for a, b in zip(vec1, vec2))
        magnitude1 = np.sqrt(sum(a * a for a in vec1))
        magnitude2 = np.sqrt(sum(b * b for b in vec2))
        
        if magnitude1 == 0 or magnitude2 == 0:
            return 0.0
        
        return dot_product / (magnitude1 * magnitude2)
    
    async def export(self, output_path: str) -> str:
        """
        Export vector store data.
        
        Args:
            output_path: Path to export data
            
        Returns:
            Path to exported data
        """
        try:
            stats = await self.get_stats()
            export_data = {
                "stats": stats,
                "index": self.index,
                "exported_at": datetime.now().isoformat()
            }
            
            with open(output_path, 'w', encoding='utf-8') as f:
                json.dump(export_data, f, ensure_ascii=False, indent=2)
            
            logger.info(f"Vector store exported to {output_path}")
            return output_path
        except Exception as error:
            logger.error("Error exporting vector store", {"error": str(error)})
            raise

File: src/rag/vectorStore.js
/**
 * Vector Store
 * Manages storage and retrieval of document embeddings
 */

const logger = require('../utils/logger');
const fs = require('fs').promises;
const path = require('path');

class VectorStore {
  constructor(config) {
    this.config = config;
    this.storePath = path.join(config.vectorStorePath, 'vectors');
    this.indexPath = path.join(config.vectorStorePath, 'index.json');
    this.maxDocuments = config.maxDocuments || 1000;
    this.documents = [];
    this.index = {};
  }

  /**
   * Initialize vector store
   */
  async initialize() {
    try {
      await fs.mkdir(this.storePath, { recursive: true });
      await this.loadIndex();
      logger.info('Vector store initialized');
    } catch (error) {
      logger.error('Error initializing vector store', { error: error.message });
      throw error;
    }
  }

  /**
   * Store document chunks with embeddings
   */
  async store(chunks) {
    try {
      await this.initialize();

      const documentId = `doc-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
      const docPath = path.join(this.storePath, `${documentId}.json`);

      const document = {
        id: documentId,
        chunks: chunks.map(chunk => ({
          id: chunk.id,
          content: chunk.content,
          embedding: chunk.embedding,
          metadata: chunk.metadata
        })),
        timestamp: new Date().toISOString(),
        chunkCount: chunks.length
      };

      // Save document to disk
      await fs.writeFile(docPath, JSON.stringify(document, null, 2));

      // Update index
      this.index[documentId] = {
        id: documentId,
        path: docPath,
        timestamp: document.timestamp,
        chunkCount: document.chunkCount,
        source: chunks[0]?.source
      };

      await this.saveIndex();

      logger.info(`Document stored successfully`, {
        documentId,
        chunkCount: chunks.length
      });

      return { id: documentId, chunkCount: chunks.length };
    } catch (error) {
      logger.error('Error storing document', { error: error.message });
      throw error;
    }
  }

  /**
   * Search for similar documents
   */
  async search(queryEmbedding, topK = 5, filter = {}) {
    try {
      await this.initialize();

      const results = [];
      const documentIds = Object.keys(this.index);

      for (const docId of documentIds) {
        const doc = await this.loadDocument(docId);
        
        if (!doc) continue;

        for (const chunk of doc.chunks) {
          // Apply filters
          if (filter.type && chunk.metadata.type !== filter.type) continue;
          if (filter.source && chunk.metadata.source !== filter.source) continue;

          // Calculate similarity
          const similarity = this.calculateCosineSimilarity(
            queryEmbedding,
            chunk.embedding
          );

          results.push({
            documentId: docId,
            chunkId: chunk.id,
            content: chunk.content,
            metadata: chunk.metadata,
            similarity
          });
        }
      }

      // Sort by similarity and return top K
      results.sort((a, b) => b.similarity - a.similarity);
      return results.slice(0, topK);
    } catch (error) {
      logger.error('Error searching vector store', { error: error.message });
      throw error;
    }
  }

  /**
   * Load document from disk
   */
  async loadDocument(documentId) {
    try {
      const docInfo = this.index[documentId];
      if (!docInfo) {
        logger.warn(`Document not found in index: ${documentId}`);
        return null;
      }

      const data = await fs.readFile(docInfo.path, 'utf8');
      return JSON.parse(data);
    } catch (error) {
      logger.error('Error loading document', { 
        documentId, 
        error: error.message 
      });
      return null;
    }
  }

  /**
   * Delete document from store
   */
  async deleteDocument(documentId) {
    try {
      const docInfo = this.index[documentId];
      if (!docInfo) {
        logger.warn(`Document not found: ${documentId}`);
        return false;
      }

      // Delete file
      await fs.unlink(docInfo.path);

      // Remove from index
      delete this.index[documentId];
      await this.saveIndex();

      logger.info(`Document deleted: ${documentId}`);
      return true;
    } catch (error) {
      logger.error('Error deleting document', { 
        documentId, 
        error: error.message 
      });
      throw error;
    }
  }

  /**
   * Save index to disk
   */
  async saveIndex() {
    try {
      await fs.writeFile(
        this.indexPath,
        JSON.stringify(this.index, null, 2),
        'utf8'
      );
    } catch (error) {
      logger.error('Error saving index', { error: error.message });
      throw error;
    }
  }

  /**
   * Load index from disk
   */
  async loadIndex() {
    try {
      const data = await fs.readFile(this.indexPath, 'utf8');
      this.index = JSON.parse(data);
      logger.info(`Loaded ${Object.keys(this.index).length} documents from index`);
    } catch (error) {
      // Index doesn't exist yet, create empty index
      this.index = {};
      await this.saveIndex();
    }
  }

  /**
   * Get store statistics
   */
  async getStats() {
    await this.initialize();

    const documentIds = Object.keys(this.index);
    let totalChunks = 0;

    for (const docId of documentIds) {
      totalChunks += this.index[docId].chunkCount;
    }

    return {
      totalDocuments: documentIds.length,
      totalChunks,
      maxDocuments: this.maxDocuments,
      availableSpace: this.maxDocuments - documentIds.length
    };
  }

  /**
   * Clear all documents from store
   */
  async clear() {
    try {
      const documentIds = Object.keys(this.index);

      for (const docId of documentIds) {
        await this.deleteDocument(docId);
      }

      logger.info('Vector store cleared');
    } catch (error) {
      logger.error('Error clearing vector store', { error: error.message });
      throw error;
    }
  }

  /**
   * Calculate cosine similarity
   */
  calculateCosineSimilarity(vec1, vec2) {
    if (vec1.length !== vec2.length) {
      throw new Error('Vector dimensions must match');
    }

    let dotProduct = 0;
    let magnitude1 = 0;
    let magnitude2 = 0;

    for (let i = 0; i < vec1.length; i++) {
      dotProduct += vec1[i] * vec2[i];
      magnitude1 += vec1[i] * vec1[i];
      magnitude2 += vec2[i] * vec2[i];
    }

    magnitude1 = Math.sqrt(magnitude1);
    magnitude2 = Math.sqrt(magnitude2);

    if (magnitude1 === 0 || magnitude2 === 0) {
      return 0;
    }

    return dotProduct / (magnitude1 * magnitude2);
  }

  /**
   * Export vector store data
   */
  async export(outputPath) {
    try {
      const stats = await this.getStats();
      const exportData = {
        stats,
        index: this.index,
        exportedAt: new Date().toISOString()
      };

      await fs.writeFile(outputPath, JSON.stringify(exportData, null, 2));
      logger.info(`Vector store exported to ${outputPath}`);
      return outputPath;
    } catch (error) {
      logger.error('Error exporting vector store', { error: error.message });
      throw error;
    }
  }
}

module.exports = VectorStore;
